src/workerd/jsg/ser.h

```diff
@@ -13,18 +13,83 @@ namespace workerd::jsg {
+//   types. This can be any enum, but it is suggested that all types in the application use the
+//   same enum type, and the numeric values of the tags must never change. A Cap'n Proto enum is
+//   suggested.
+// * `JSG_SERIALIZABLE`'s parameters are a list of tags that encode this type. There may be more
+//   than one tag listed to support versioning. The first tag listed is the current version, which
+//   is the format that `serialize()` will write. The other tags are non-current versions that
+//   `deserialize()` accepts in addition to the current version. These are usually old versions,
+//   but could also include a new version that hasn't fully rolled out yet -- it will be necessary
+//   to fully roll out support for parsing a new version before anyone can start generating it.
+// * The serialization system automatically handles writing and reading the tag values before
+//   calling the methods.
+// * serialize() makes a series of calls to serializer.write*() methods to write the content of the
+//   object.
+// * deserialize() makes a corresponding series of calls to deserializer.read*() methods to read
+//   the object. These must be the exact corresponding calls in the same order as serialize()
+//   would have made them. The sequence can never change once data has been written for a given
+//   tag version; the only way to change is to define a new version.
 class Serializer final: v8::ValueSerializer::Delegate {
 public:
+  // "Externals" are values which can be serialized, but refer to some external resource, rather
+  // than being self-contained. The way externals are supported depends on the serialization
+  // context: passing externals over RPC, for example, is completely different from storing them
+  // to disk.
+  //
+  // A `Serializer` instance may have an `ExternalHandler` which can be used when serializing
+  // externals. This type has no methods, but is meant to be subclassed. A host object which
+  // represents an external and is trying to serialize itself should use dynamic_cast to try to
+  // downcast `ExternalHandler` to any particular handler interface it supports. If the handler
+  // doesn't implement any supported subclass, then serialization is not possible, and an
+  // appropriate exception should be thrown.
+  class ExternalHandler {
+  public:
+    // We declare the destructor just so that this class has a virtual method, which ensures it is
+    // polymorphic (has a vtable) so that dynamic_cast can be used on it. We mark this method pure
+    // (`= 0`) so that `ExternalHandler` itself cannot be instantiated.
+    virtual ~ExternalHandler() noexcept(false) = 0;
+  };
+
   struct Options {
     // When set, overrides the default wire format version with the one provided.
     kj::Maybe<uint32_t> version;
     // When set to true, the serialization header is not written to the output buffer.
     bool omitHeader = false;
+
+    kj::Maybe<ExternalHandler&> externalHandler;
```

Comment:
nit: a short note here referencing where the ownership of the `ExternalHandler&` ref is expected to be held would be helpful.

Reply:
f1ee030f